### PR TITLE
Implement escape scopes for all types of local symbols

### LINKED ISSUE #22206: EE escape analysis crashes when compiling lambdas
Code:

```csharp
namespace ConsoleApp3
{
    public delegate ref  int D();

    class Program
    {
        static void Main(string[] args)
        {
            int x = 0;
        }

        static ref  int M(D lambda)
        {
            return ref lambda();
        }
    }
}
```

Put a breakpoint after `int x = 0` in `Main()`, and F5. When it hits, evaluate the expression `M(() => ref x)`.

**Expectation:**  error CS8168: Cannot return local 'x' by reference because it is not a ref local
**Actual:** compiler crashes because `RefEscapeScope` is not implemented for EE symbols.

**Suggestion**: Both `RefEscapeScope` and `ValEscapeScope` should be abstracted and implemented wherever possible, instead of throwing an unreachable exception.

### PR DESCRIPTION
**Pull request by OmarTawfik**
Fixes #22206

cc @dotnet/roslyn-compiler: @cston @VSadov @cston for code review.
cc @jaredpar for ask mode approval.

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@tmat @cston AFAIK, there is no infrastructure to test evaluating expressions in the EE and getting their result (a unit test, not an E2E scenario test). Did I miss something?
The current EE tests I found test that things get compiled, but don't actually execute expressions at runtime (through the debugger) and get their values.

**Comment by tmat** [comment]
@OmarTawfik That's correct. For Roslyn such test would be an integration test as you'd need to spin up VS, create a project, launch it and attach to the process. Why do you need that?

**Comment by OmarTawfik** [comment]
@tmat I don't think this bug belongs to this class of tests. Just wanted to make sure I'm not missing something. Thanks!

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler another sign off?
@jaredpar for ask mode approval.

File: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs [hunk 1]
```diff
@@ -2156,5 +2156,17 @@ internal override RefKind RefKind
         {
             get { return RefKind.None; }
         }
+
+        /// <summary>
+        /// Compiler should always be synthesizing locals with correct escape semantics.
+        /// Checking scape scopes is not valid here.
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs]: Nit: typo on escape

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/TypeSubstitutedLocalSymbol.cs [hunk 2]
```diff
@@ -93,6 +93,10 @@ internal override RefKind RefKind
             get { return _originalVariable.RefKind; }
         }
 
+        internal override uint ValEscapeScope => _originalVariable.ValEscapeScope;
+
+        internal override uint RefEscapeScope => _originalVariable.RefEscapeScope;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/TypeSubstitutedLocalSymbol.cs]: Is this possible to cover this with a test?

File: src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EELocalSymbolBase.cs [hunk 3]
```diff
@@ -76,5 +76,17 @@ internal sealed override DiagnosticInfo GetUseSiteDiagnostic()
...
+        /// EE Symbols have no source symbols associated with them.
+        /// They should be safe to escape for evaluation purposes.
+        /// </summary>
+        internal override uint ValEscapeScope => Binder.TopLevelScope;
+
+        /// <summary>
+        /// EE Symbols have no source symbols associated with them.
+        /// They should be safe to escape for evaluation purposes.
+        /// </summary>
+        internal override uint RefEscapeScope => Binder.TopLevelScope;
```
Comment (jcouv) [inline src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EELocalSymbolBase.cs]: Is this testable?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/TypeSubstitutedLocalSymbol.cs (hunk 2 above)
Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/TypeSubstitutedLocalSymbol.cs]: Actually, will replace that with a throw as well. Compiler shouldn't do that for synthesized symbols at all.

Comment (OmarTawfik) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/TypeSubstitutedLocalSymbol.cs]: This is only used during lowering. All escape errors should be reported before that.

**Comment by OmarTawfik** [comment]
cc @jaredpar for ask mode approval.

**Comment by MeiChin-Tsai** [comment]
escrow approved by Mei-Chin.

**Comment by cston** [comment]
Please add a test, such as:
```
        [Fact]
        public void RefReturnNonRefLocal()
        {
            var source =
@"public delegate ref int D();
class C
{
    static void Main()
    {
        int x = 0;
    }
    static ref int M(D d)
    {
        return ref d();
    }
}";
            var comp = CreateStandardCompilation(source, options: TestOptions.DebugDll);
            WithRuntimeInstance(comp, runtime =>
            {
                var context = CreateMethodContext(runtime, "C.Main");
                string error;
                context.CompileExpression("M(() => ref x)", out error);
                Assert.Null(error);
            });
        }
```

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs [hunk 4]
```diff
@@ -6382,5 +6382,30 @@ void M(D lambda)
             Assert.Contains(AttributeDescription.CodeAnalysisEmbeddedAttribute.FullName + "..ctor()", methodsGenerated);
             Assert.Contains(AttributeDescription.IsReadOnlyAttribute.FullName + "..ctor()", methodsGenerated);
         }
+
+        [Fact]
+        public void RefReturnNonRefLocal()
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/ExpressionCompilerTests.cs]: Please add `[WorkItem(22206, "https://github/...")]`

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-infrastructure interminnent [failure](https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_release_unit32_prtest/10294/testReport/junit/Microsoft.CodeAnalysis.Editor.UnitTests.CodeGeneration/CodeGenerationTests+VisualBasic/AddMustOverrideMethod/)

**Comment by natidea** [comment]
Shiproom approved to merge via [510393](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/510393)

**Pull request merged**